internal/watermill/middleware/dlq.go

```diff
@@ -0,0 +1,96 @@
+package middleware
+
+import (
+	"fmt"
+
+	"github.com/ThreeDotsLabs/watermill/message"
+	multierror "github.com/hashicorp/go-multierror"
+	"github.com/pkg/errors"
+
+	"github.com/openmeterio/openmeter/internal/watermill"
+)
+
+const (
+	DLQErrorReasonKey   = "reason_dlq"
+	DLQOriginalTopicKey = "original_topic"
+)
+
+// DLQMsgProcessorFunc determines whether any messages should be published to the DLQ based on the original message and the error.
+// Even if DLQMsgProcessorFunc returns a non-nil slice of messages, if error is nil no messages will be published to the DLQ.
+type DLQMsgProcessorFunc func(original *message.Message, err error) []*message.Message
+
+// defaultProcessor listens for RetryableErrors and publishes the messages contained in that
+var defaultProcessor DLQMsgProcessorFunc = func(original *message.Message, err error) []*message.Message {
+	if retryable, ok := err.(watermill.RetryableError); ok {
+		return retryable.RetryMessages()
+	}
+
+	return nil
+}
+
+type deadLetterQueue struct {
+	topic string
+	pub   message.Publisher
+
+	msgProcessor DLQMsgProcessorFunc
+}
+
+// DLQ provides a middleware that salvages unprocessable messages and published them on a separate topic.
+// The main middleware chain then continues on, business as usual.
+func DLQ(pub message.Publisher, topic string, msgProcessor DLQMsgProcessorFunc) (message.HandlerMiddleware, error) {
+	if topic == "" {
+		return nil, fmt.Errorf("invalid DLQ topic")
+	}
+
+	if msgProcessor == nil {
+		msgProcessor = defaultProcessor
+	}
+
+	dlq := deadLetterQueue{
+		topic:        topic,
+		pub:          pub,
+		msgProcessor: msgProcessor,
+	}
+
+	return dlq.Middleware, nil
+}
+
+func (dlq deadLetterQueue) publishWithErr(msg *message.Message, err error) error {
+	// Sanity check, we couldn't annotate the metadata without the error
+	if err == nil {
+		return nil
+	}
+
+	// Add context why it failed
+	msg.Metadata.Set(DLQErrorReasonKey, err.Error())
+	msg.Metadata.Set(DLQOriginalTopicKey, message.SubscribeTopicFromCtx(msg.Context()))
+
+	// Don't intercept error from publish. Can't help you if the publisher is down as well.
+	return dlq.pub.Publish(dlq.topic, msg)
+}
+
+func (dlq deadLetterQueue) Middleware(h message.HandlerFunc) message.HandlerFunc {
+	// Capture return values from the handler
+	return func(msg *message.Message) (events []*message.Message, err error) {
```

Comment:
Please don't use named returns: Named returns only make sense if you are trying to manipulate return values in a defer when using recover, but panic handling is handled by other middleware.

Please don't use defer as this code can be replaced by:

```
func(msg * ...)
  outMessages, err := h(msg)
  
 // code in defer
```

Reply:
I'm OK with the change, but thats a stylistic preference / convention

(technically the two [arent't the same](https://go.dev/play/p/_LksNUvLsOl) but we don't need that here (nor can we use it due to the nil check a few lines up in `publishWithErr`)